prober/icmp.go

```diff
@@ -234,15 +234,31 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 	rttStart := time.Now()
 
 	if icmpConn != nil {
+		ttl := module.ICMP.TTL
+		if ttl > 0 {
+			if c4 := icmpConn.IPv4PacketConn(); c4 != nil {
+				level.Debug(logger).Log("msg", "Setting TTL (IPv4 unprivileged)", "ttl", ttl)
+				c4.SetTTL(ttl)
+			}
+			if c6 := icmpConn.IPv6PacketConn(); c6 != nil {
+				level.Debug(logger).Log("msg", "Setting TTL (IPv6 unprivileged)", "ttl", ttl)
+				c6.SetHopLimit(ttl)
+			}
+		}
 		_, err = icmpConn.WriteTo(wb, dst)
```

Comment:
This one is the one that has me scratching my head...

What happens here if module.ICMP.TTL is zero? It uses the operating system's current setting. On Linux that's usually 64 (but it's distribution dependent). I have no idea what it is elsewhere.

Reply:
Yeah, this had me scratching my head too.

Unless I screwed up, this should preserve the existing behaviour:
- If we're using the unprivileged mode, it defaults to the OS's TTL, I believe, because it doesn't explicitly set the TTL: https://github.com/prometheus/blackbox_exporter/blob/f0d1414f1fa6fd0b0c4052e4dfae9b9e721060bd/prober/icmp.go#L237
- If we're using the IPv4 raw mode, it uses the default baked into the blackbox_exporter (i.e. 64): https://github.com/prometheus/blackbox_exporter/blob/f0d1414f1fa6fd0b0c4052e4dfae9b9e721060bd/prober/icmp.go#L245

I'm not sure if this difference in behaviour was by design. I am okay to make the changes in this PR so that the behaviour is more consistent, e.g. use the default value baked into blackbox_exporter.

With apologies, it's a new experience for me to contribute to this project so I'm not sure who can/should make these decisions (or even who can accept this PR). I am happy to make appropriate changes if someone points me in the right direction.

Reply:
That's alright.

Because of the way this is implemented, I doubt many people are relying on falling back to the OS setting.

Let's go forward with this as is.